File: bot.py
import datetime
import discord
import os
import random
from discord.ext import tasks
from dotenv import load_dotenv
from zoneinfo import ZoneInfo

load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
CHANNEL_ID = 1037366923324838008

intents = discord.Intents.default()
intents.message_content = True

client = discord.Client(intents=intents)

midnight_time = datetime.time(hour=0, minute=0, second=0, tzinfo=ZoneInfo("Asia/Shanghai"))
noon_time = datetime.time(hour=24, minute=0, second=0, tzinfo=ZoneInfo("Asia/Shanghai"))

all_pins = []


# get all pins
async def get_pins():
    print("Attempting to retrieve all pins.")

    all_channels = []
    for guild in client.guilds:
        for channel in guild.text_channels:
            all_channels.append(channel)

    for channel in all_channels:
        all_pins.extend(await channel.pins())

    print("Successfully retrieved all pins.")


@client.event
async def on_ready():
    await get_pins()

    # start daily wisdom on ready
    if not midnight_wisdom.is_running():
        midnight_wisdom.start()
        print("Midnight wisdom task started.")
    if not noon_wisdom.is_running():
        noon_wisdom.start()
        print("Noon wisdom task started.")


async def bestow_wisdom(channel):
    response = random.choice(all_pins)
    content_format = ""
    if response.content:
        content_format = ' "' + response.content + '"'
    content_format.replace('@', '@ ')
    await channel.send('Confucius say: ' + content_format, files=[await f.to_file() for f in response.attachments])
    await channel.send(response.jump_url)


# on message commands
@client.event
async def on_message(message):
    if message.author == client.user:
        return

    if message.content.lower() == 'confucius bestow upon me your wisdom':
        await bestow_wisdom(message.channel)

    elif message.content.lower() == 'confucius how big is your brain':
        await message.channel.send('As large as ' + str(len(all_pins)) + ' pins.')


# midnight wisdom
@tasks.loop(time=midnight_time)
async def midnight_wisdom():
    print("Attempting to send midnight wisdom.")
    channel = client.get_channel(1037366923324838008)

    await bestow_wisdom(channel)

    print("Midnight wisdom sent.")


# noon wisdom
@tasks.loop(time=noon_time)
async def noon_wisdom():
    print("Attempting to send noon wisdom.")
    channel = client.get_channel(1037366923324838008)

    await bestow_wisdom(channel)

    print("Noon wisdom sent.")


# pin updates
@client.event
async def on_guild_channel_pins_update(channel, last_pin):
    await get_pins()


@client.event
async def on_guild_channel_pins_update(channel, last_pin):
    await get_pins()


client.run(TOKEN)

File: bot.js
const { Client, GatewayIntentBits, Collection } = require("discord.js");
const { config } = require("dotenv");
const cron = require("node-cron");

config();

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
  ],
});

let allPins = new Collection();

// Fetch all pins from all text channels in all guilds
async function getPins() {
  console.log("Attempting to retrieve all pins.");
  allPins.clear();

  for (const guild of client.guilds.cache.values()) {
    for (const channel of guild.channels.cache.values()) {
      if (channel.isTextBased()) {
        try {
          const pins = await channel.messages.fetchPinned();
          pins.forEach((pin) => allPins.set(pin.id, pin));
        } catch (error) {
          console.error(
            `Failed to retrieve pins from ${channel.name}: ${error.message}`
          );
        }
      }
    }
  }
  console.log("Successfully retrieved all pins.");
}

// When the bot is ready
client.once("ready", async () => {
  try {
    await getPins();

    // Schedule tasks for midnight and noon
    cron.schedule("0 0 * * *", () => bestowWisdom(process.env.CHANNEL_ID), {
      timezone: "Asia/Shanghai",
    });

    console.log("Bot is ready and scheduled tasks are set.");
  } catch (error) {
    console.error("Error during bot initialization:", error.message);
  }
});

// Send a random pinned message to the channel
async function bestowWisdom(channelId) {
  try {
    const channel = await client.channels.fetch(channelId);
    if (!channel) {
      console.error(`Channel with ID ${channelId} not found.`);
      return;
    }

    if (allPins.size === 0) {
      await getPins();
    }

    const randomPin = allPins.random();
    if (!randomPin) {
      console.log("No pinned messages available to send.");
      return;
    }

    // If strict, we replace the word directly regardless of text before or after
    // If not strict, there needs to be a word boundary before or after,
    // since short words like "sha" may be contained within other words
    const wordReplacements = {
      shaheen: {
        replacement: "sora",
        strict: true,
      },
      sha: {
        replacement: "sora",
        strict: false,
      },
    };

    let messageToSend = randomPin.content ? `"${randomPin.content}"` : "";

    for (const [key, value] of Object.entries(wordReplacements)) {
      if (value.strict) {
        const regex = new RegExp(key, "gi");
        messageToSend = messageToSend.replace(regex, value.replacement);
      } else {
        const regex = new RegExp(`\\b${key}\\b`, "gi");
        messageToSend = messageToSend.replace(regex, value.replacement);
      }
    }

    let messageContent = `Confucius say: ${messageToSend}\n${randomPin.url}`;

    const messageOptions = {
      content: messageContent,
      files:
        randomPin.attachments.size > 0
          ? [...randomPin.attachments.values()].map(
              (attachment) => attachment.url
            )
          : [],
    };

    await channel.send(messageOptions);
    console.log("Successfully sent wisdom to channel.");
  } catch (error) {
    if (error.code === 50013) {
      console.error(
        `Missing permissions to send message to channel ${channelId}. Error: ${error.message}`
      );
    } else if (error.code === 50001) {
      console.error(
        `Channel ${channelId} not found or bot doesn't have access. Error: ${error.message}`
      );
    } else if (error.code === 50005) {
      console.error(
        `Cannot send empty message to channel ${channelId}. Error: ${error.message}`
      );
    } else {
      console.error(
        `Error sending wisdom to channel ${channelId}: ${error.message}`
      );
    }
  }
}

// Handle messages for specific commands
client.on("messageCreate", async (message) => {
  if (message.author.bot) return;

  try {
    if (
      message.content.toLowerCase() === "confucius bestow upon me your wisdom"
    ) {
      await bestowWisdom(message.channel.id);
    } else if (
      message.content.toLowerCase() === "confucius how big is your brain"
    ) {
      await message.channel.send(`As large as ${allPins.size} pins.`);
    }
  } catch (error) {
    console.error(`Error handling message command: ${error.message}`);
    try {
      await message.channel.send(
        "I apologize, but I encountered an error while processing your request."
      );
    } catch (sendError) {
      console.error(`Failed to send error message: ${sendError.message}`);
    }
  }
});

// Update pins whenever a new pin is added
client.on("channelPinsUpdate", async (channel, time) => {
  try {
    await getPins();
  } catch (error) {
    console.error(`Error updating pins: ${error.message}`);
  }
});

// Handle client errors
client.on("error", (error) => {
  console.error("Discord client error:", error.message);
});

// Handle process errors
process.on("unhandledRejection", (reason, promise) => {
  console.error("Unhandled Rejection at:", promise, "reason:", reason);
});

process.on("uncaughtException", (error) => {
  console.error("Uncaught Exception:", error.message);
  console.error("Stack trace:", error.stack);
});

client.login(process.env.DISCORD_TOKEN);
